cpp/src/buffer/buffer.cpp

```diff
@@ -139,6 +145,133 @@ std::unique_ptr<Buffer> Buffer::copy(MemoryType target, rmm::cuda_stream_view st
     );
 }
 
+std::unique_ptr<Buffer> Buffer::copy_slice(
+    std::ptrdiff_t offset, std::ptrdiff_t length, rmm::cuda_stream_view stream
+) const {
+    RAPIDSMPF_EXPECTS(offset <= std::ptrdiff_t(size), "offset can't be more than size");
+    RAPIDSMPF_EXPECTS(
+        offset + length <= std::ptrdiff_t(size), "offset + length can't be more than size"
+    );
+    return std::visit(
+        overloaded{
+            [&](const HostStorageT& storage) {  // host -> host
+                auto host_buf = std::unique_ptr<Buffer>(new Buffer{
+                    std::make_unique<std::vector<uint8_t>>(
+                        storage->begin() + offset, storage->begin() + offset + length
+                    ),
+                    br
+                });
+                host_buf->override_event(event_);  // if there was an event, use it
```

Comment:
Can you give `event_` to the Buffer's ctor instead?

Reply:
@madsbk HostBuffer constructor doesnt have an event arg :slightly_frowning_face:

Reply:
So what happens if the data in `this` buffer hasn't been written yet? Since we do not wait on `event_`, what will happen in:
```c++
                   std::make_unique<std::vector<uint8_t>>(
                        storage->begin() + offset, storage->begin() + offset + length
                    ),
```

Reply:
@madsbk Ah! that's a good point. `throw` if its not ready, maybe? or return `nullptr` so that it can be attempted later on?

Reply:
Or wait on the event, `cudaEventSynchronize` ?

Reply:
True, but if the alternative is a crash? Maybe we should have wrapper function of `cudaEventSynchronize` that times the call and adds it to the statistics. My guess is that we almost never would block here.

Reply:
I think that's true. Okay, let me add that.

Reply:
@madsbk I have added the changes now